image/templates/helm/shared/templates/_scanner-v4_init.tpl.htpl

```diff
@@ -0,0 +1,108 @@
+  running in Matcher mode. In this context, the Helm chart can create its own certificates.
+
+  StackRox's Secured Cluster services may deploy the Scanner in Indexer mode, only.
+  This would be done to access registries inaccessible to the Central cluster.
+  In this context, the Helm chart does not generate its own certificates.
+
+  $scannerV4Config contains all values which are configured by the user. The structures can be viewed in the respective
+  config-shape. See internal/scanner-v4-config-shape.yaml.
+   */}}
+
+{{ define "srox.scannerV4Init" }}
+
+{{ $ := index . 0 }}
+{{ $scannerV4Cfg := index . 1 }}
+{{ $_ := false }}
+
+{{/* Handle high-level defaulting logic. */}}
+{{/* Treat a scannerV4.disable=false setting as a hint to enable indexer and/or matcher depending on the chart being used. */}}
+{{- if and (not $scannerV4Cfg.disable) (kindIs "invalid" $scannerV4Cfg.indexer.disable) (kindIs "invalid" $scannerV4Cfg.matcher.disable) -}}
+{{- /* The top-level setting scannerV4.disable=false activates a default setting based on the chart. */ -}}
+{{- if eq $.Chart.Name "stackrox-central-services" -}}
+  {{- $_ = set $scannerV4Cfg.indexer "disable" false -}}
+  {{- $_ = set $scannerV4Cfg.matcher "disable" false -}}
+{{- else if eq $.Chart.Name "stackrox-secured-cluster-services" -}}
+  {{- $_ = set $scannerV4Cfg.indexer "disable" false -}}
+{{- else -}}
+  {{- include "srox.fail" (printf "Unexpected Helm chart name %q." $.Chart.Name) -}}
+{{- end -}}
+{{- else -}}
+{{- end -}}
+
+{{/* Sanity checks. */}}
+{{- if eq $.Chart.Name "stackrox-central-services" -}}
+  {{- if and (not $scannerV4Cfg.indexer.disable) $scannerV4Cfg.matcher.disable -}}
+    {{- include "srox.fail" "This chart only supports deploying indexer and matcher together." -}}
+  {{- end -}}
+  {{- if and (not $scannerV4Cfg.matcher.disable) $scannerV4Cfg.indexer.disable -}}
+    {{- include "srox.fail" "This chart only supports deploying indexer and matcher together." -}}
+  {{- end -}}
+{{- end -}}
+{{- if eq $.Chart.Name "stackrox-secured-cluster-services" -}}
+  {{- if not $scannerV4Cfg.matcher.disable -}}
+    {{- include "srox.fail" "This chart only supports deploying the scanner V4 indexer." -}}
+  {{- end -}}
+{{- end -}}
+
+{{/* Provide some human-readable feedback regarding the Scanner V4 configuration to the user installing the Helm chart. */}}
+{{- $scannerV4Mode := "none" -}}
+{{- if and (not $scannerV4Cfg.indexer.disable) (not $._rox.scannerV4.matcher.disable) -}}
+{{- $scannerV4Mode = "indexer-and-matcher" -}} 
+{{- else if not $scannerV4Cfg.indexer.disable -}}
+{{- $scannerV4Mode = "indexer-only" -}}
+{{- else if not $scannerV4Cfg.matcher.disable -}}
+{{- $scannerV4Mode = "matcher-only" -}}
+{{- end -}}
+
+{{- $indexerEnabled := false -}}
+{{- $matcherEnabled := false -}}
+{{- $dbEnabled := false -}}
+
+{{ if and (not (kindIs "invalid" $scannerV4Cfg.indexer.disable)) (not $scannerV4Cfg.indexer.disable) }}
```

Comment:
It's very hard for me to understand the entire enabled/disabled logic here. One of the reason is that we have the flip from first all field are `disabled` but after this init the have `enabled` fields. I imagine it could be simpler if we have `enabled` in the config in the first place.

Also IIUC how it should work:
- `scannerV4.disabled == true` set matcherEnabled, indexerEnabled and dbEnabled to false
- `scannerv4.disabled == false`
  - always set dbEnabled == true
  - if `chart == central-services` set matcherEnabled and indexerEnabled to true
  - if `chart == secured-cluster` set matcherEnabled to false and indexerEnabled to true

This looks way simpler than the current code. Also we wouldn't need the `indexer.disabled`and `matcher.disabled` fields at all, since this is driven by the chart name instead.

Reply:
Is way better to understand now, thanks.